### PR TITLE
Quick cache campaign™

### PR DESCRIPTION
**Pull request by mendelB**
### What does this PR do?
- Implements a bare-bones caching strategy for Campaigns to prevent a trazillion calls to Contentful API

### Any background context you want to provide?
Using the `Cache` class in the `CampaignRepository` repository to stash the campaign for 15-minute increments.
The annoying thing was that once you serialize the `campaign` data, contentful cries when you try and create a `Campaign` entity with that data, for it requires a `DynamicEntry` argument to the constructor. 
A little 🔍 revealed a `reviveJson` method in our Contentful API client, allowing us to use the cached JSON to create a campaign and help make the world a better place!

### What are the relevant tickets/cards?
https://www.pivotaltracker.com/story/show/152202846

### Checklist
- [ ] Documentation added for new features/changed endpoints.
- [ ] Added appropriate feature/unit tests.

### CODE REVIEW

File: app/Repositories/CampaignRepository.php [hunk 1]
```diff
@@ -45,19 +48,37 @@ public function getAll()
...
     {
-        $query = (new Query)
+        if (Cache::has($slug)) {
+            $cachedCampaignJson = Cache::get($slug);
+
+            // This turns the JSON into a Contentful Dynamic Entry
+            // so we can create a campaign entity out of the cached data
+            $revivedContentfulEntry = $this->client->reviveJson($cachedCampaignJson);
+
+            $campaign = new Campaign($revivedContentfulEntry);
```
Comment (weerd) [inline app/Repositories/CampaignRepository.php]: Line 58 and 78 are pretty much the same. We can likely take them out of the conditional and place right before the `return` call, and then just rename the var to something that's the same in both conditions... I'll roll over IRL :)

File: app/Repositories/CampaignRepository.php [hunk 2]
```diff
@@ -45,19 +48,35 @@ public function getAll()
...
-        $campaigns = $this->makeRequest($query);
+            $campaigns = $this->makeRequest($query);
+
+            if (! $campaigns->count()) {
+                throw new ModelNotFoundException;
+            }
+
+            $campaignEntry = $campaigns[0];
+
+            if (! App::environment(['local', 'staging'])) {
```
Comment (weerd) [inline app/Repositories/CampaignRepository.php]: @mendelB @itsjoekent I'm debating if we would prefer to go by our environments, or potentially set a new env variable like `APP_CACHE=true` so it's something we can easily toggle on/off if we _want_ to see caching working locally or on staging etc... thoughts?

Comment (itsjoekent) [reply app/Repositories/CampaignRepository.php]: Yeah I think that could be fine as well

Comment (mendelB) [reply app/Repositories/CampaignRepository.php]: Cool yeah that feels like a good idea!

Comment (mendelB) [reply app/Repositories/CampaignRepository.php]: @weerd So we'd be assuming that the `APP_CACHE` is set to `true` or `false` and basing caching on that? meaning if the variable does *not* exist it wouldn't cache?

**Comment by itsjoekent** [comment]
@mendelB Can you use the config/services file https://github.com/DoSomething/phoenix-next/blob/79ca93cc5b40e8587e995e7cf29aaa8e3fb1f37c/config/services.php#L19 and reference it with the config function? https://github.com/DoSomething/phoenix-next/blob/cefffd6df344e1fe70f62e82acab44592e379b26/app/helpers.php#L270

File: app/Repositories/CampaignRepository.php [hunk 3]
```diff
@@ -45,19 +47,35 @@ public function getAll()
...
-        $campaigns = $this->makeRequest($query);
+            $campaigns = $this->makeRequest($query);
+
+            if (! $campaigns->count()) {
+                throw new ModelNotFoundException;
+            }
+
+            $campaignEntry = $campaigns[0];
+
+            if (env('CONTENTFUL_CACHE')) {
```
Comment (weerd) [inline app/Repositories/CampaignRepository.php]: Yeah use `config()` instead since that uses the cached config info

**Comment by mendelB** [comment]
Which function in `helpers.php` would I be adding this to? Or create a new function there for this?

**Comment by itsjoekent** [comment]
@mendelB Was just showing an example of the `config` function

**Comment by mendelB** [comment]
Gotcha gotcha! was confused for a sec

File: config/services.php [hunk 4]
```diff
@@ -55,4 +55,8 @@
         'prefix' => env('SIXPACK_COOKIE_PREFIX'),
         'timeout' => env('SIXPACK_TIMEOUT'),
     ],
+
+    'contentful' => [
+        'cache' => env('CONTENTFUL_CACHE'),
```
Comment (weerd) [inline config/services.php]: Maybe default this to `true`.

**Comment by mendelB** [comment]
@weerd Should it also be set to `true` in the .env example?

**Pull request merged**